### ISSUE TITLE
AudioOutStream.FlushAsync hung indefinitely under certain edge case conditions

### ISSUE THREAD

**Issue by Inirit** [RogueException/Discord.Net#687]
**tl;dr;**
Flushing an audio stream caused an indefinite hang when the audio source has a total length smaller than the individual buffer length.

**Details**
Something I've been working on recently is streaming audio data to a Discord voice channel, I had it working for some simple cases but I came across a scenario where calling FlushAsync for an AudioOutStream was resulting in the call hanging indefinitely.

I've been basing my experimenting on the guide for audio streaming written [here](https://discord.foxbot.me/docs/guides/voice/sending-voice.html), I see the warning that it's out of date so I've been cautious in my usage of it. My implementation currently looks like this (there's many engineering things wrong with this, I'm still iterating on my approach and debugging certain things):

```C#
public virtual async Task BroadcastAudio()
{
	using (var audioClient = CreateAudioClient())
	using (var sourceStream = new FileStream(FilePath, FileMode.Open))
	using (var targetStream = audioClient.CreatePCMStream(AudioApplication.Mixed, 90000))
	{
		try
		{
			await sourceStream.CopyToAsync(targetStream);

			await targetStream.FlushAsync();
		}
		catch (Exception e)
		{
			Console.WriteLine(e.ToString());
		}
	}
}
```

The CreateAudioClient function is mostly just the logic for getting the channel to be streamed to, ultimately it creates the audio client like this:
```C#
var audioClient = targetChannel.ConnectAsync().GetAwaiter().GetResult();
```

The source stream uses a "cached" file that I've already created with ffmpeg using the following command:
```
ffmpeg -i <in_file> -ac 2 -f s16le -ar 48000 <out_file>
```
I've verified that these files were created successfully and that they sound correct by opening them in Audacity.

It just so happens that some of the audio sources I want to use are only a few tenths of a second long. When I used one of these files as the source for streaming, the call to FlushAsync would hang and no audio would be played. After a bunch of debugging (I eventually just downloaded the Discord.Net source so I could step into it), I realized that the size of the buffer being used to stream the data was configured as 1 second (or 1000ms) via the default value of CreatePCMStream and that this was the cause of failure. I changed my call to CreatePCMStream by specifying a smaller buffer size, I chose 200ms somewhat arbitrarily:
```C#
var targetStream = audioClient.CreatePCMStream(AudioApplication.Mixed, 90000, 200)
```
With this change, all of my audio sources were now streaming correctly.

At first I thought maybe this was just my own failure to configure things correctly, but then I figured this likely wasn't necessarily the case. It seems like it would be unreasonable to expect an input stream's total size to be an exact multiple of the output stream's buffer size, so I'd assume there to be logic at some point to handle when an output buffer cannot be completely filled. Perhaps it's just this edge case where the very first buffer couldn't be filled that's causing the hanging.

I happen to be using 1.0.0-rc3-00746 from the Discord.Net MyGet service, but I was originally using the 1.0.0-rc2 NuGet package and was getting the same results.

**Comment by Quahu** [comment]
Set the bitrate in CreatePCMStream to 98304 which is 96 * 1024 instead of 90000.

**Comment by Inirit** [comment]
Will that have an effect on the behavior of this edge case condition? 90000 was just a constant I threw in to get it working, the guide I linked actually had 1024 listed as the value to put there but I'm assuming the underlying code for that method changed since the guide was written despite the signature of the method remaining the same.

**Comment by Joe4evr** [comment]
It's the exact same value [the lib sets as default](https://github.com/RogueException/Discord.Net/blob/dev/src/Discord.Net.WebSocket/Audio/AudioClient.cs#L162).

**Comment by Inirit** [comment]
Right, but I'm not sure how that relates to this case in particular. I can modify my call to use the default and see how that changes things once I'm able to, but in the meantime I was curious as to what effect this would have other than essentially increasing the output quality of the stream.

**Comment by khionu** [comment]
It's good practice to keep values consistent and close to multiples of powers of 2.

**Comment by Inirit** [comment]
I'd argue against it being good practice to keep it consistent when it's exposed as an argument to a public method. Being an argument implies that changes to it are supported by the underlying implementation, either through flexibility of the logic or by error handling of invalid conditions. As an external user, I wouldn't have any internal or tribal knowledge to direct my usage of it otherwise. Being a power of 2 is more reasonable due to the nature of the typical platform architecture this code would run on, but even then I'd argue against it being communicated as good practice and instead draw the distinct line between what is supported and what is not so that it's clear to any arbitrary 3rd party who intends to make use of it.

That's my personal philosophy, though, and I don't mean to impose it on this project since I'm not a contributor. I'm just reporting a case where the current implementation hangs under what seem to be valid conditions and it's unclear to me whether the suggestions in the comments are meant to be interpreted as possible solutions or are instead general critiques of the code snippet I provided.

**Comment by khionu** [comment]
I mean, the project allows for 0 as a valid value..... other values should work, but like I said, multiples of powers of 2 (ie 3096, 3 * 2^10) are best, for various reasons, not just architecture.

That said, the fact that you've narrowed it down to a specific scenario means there is likely an actual cause. I'll take a peak, when I have a chance.

**Comment by Inirit** [comment]
In practice there's definitely a level of intuition that can be reasonably expected of an external user. For example, if you passed 0 or even a negative value as a bitrate it should be obvious that it's not going to work and if I were to file a bug about how the method allows such an input I'd expect it to be punted as being too low priority against other tasks. In an ideal world that adhered to rigid standards I'd expect that the entire range of input for a public method is handled in some way or another even if it's just throwing an exception at runtime for invalid ranges, but that tends to not be a realistic goal within any reasonable time frame.

In this case, given a base intuition that the value should be positive for an otherwise unconstrained integer input, any value of integer seems like it would be valid from my external perspective. I'm not very familiar with all the intricacies that go into audio encoding/decoding and I'm sure there's real reasons why certain bit rates are preferred, perhaps those reasons should be part of that baseline intuition and I just happen to lack it. If there's a significant effect of such a configuration in Discord.Net specifically, though, I would suggest that the input either be constrained to a subset of values or to log a "warning" to the developer that the input is technically valid and will probably work but it's not ideal.

Again I don't mean to impose, I just can't help myself from arguing about things that ultimately don't matter that much. Thanks for taking a look when you can.

**Comment by BinkanSalaryman** [comment]
CreatePCMStream:
calling IAudioStream#Flush before preload fulfilled causes busy waiting in BufferedWriteStream.cs

CreateDirectPCMStream:
calling IAudioStream#Flush causes calling loop between #FlushAsync and #Flush in AudioStream.cs (OutputStream.cs)
